enhancement(tests): Kubernetes E2E test framework (#2702)
timberio/vector: CONTRIBUTING.md

Essentially, the steps you have to take to deploy manually are the same that
`skaffold` will perform, and they're outlined at the previous section.

#### Kubernetes E2E tests

> Kubernetes integration is quite complicated, a lot of things can go wrong.

Reviewer: ```suggestion
Kubernetes integration has a lot of parts that can go wrong.
```

Slight reword so we don't imply that kubernetes is too complicated for the user.

To cope with the complexity and ensure we maintain high quality, we use
E2E (end-to-end) tests.

> E2E tests normally run at CI, so there's typically no need to run them
> manually.

##### Requirements

* `kubernetes` cluster (`minikube` has special support, but any cluster should
  work)
* `docker`
* `kubectl`

> * `bash`

Reviewer: So these tests can't work from a Windows machine?

Author: Technically, you should be able to run tests on Windows via git bash. I didn't test it though...
Also, I want to note that running tests *on* Windows doesn't correlate to testing *Windows clusters*, and that we don't address that yet.

Reviewer: Yeah I have bash through busybox so I think this might work. I think this is fine for now. We'll need to support windows fully at some point though. We do support users developing on windows.

Author: > We do support users developing on windows.
Yeah, that's a bit tricky to do for k8s, but we'll get there.

The dev flow (the one that uses skaffold) *is* currently limited to linux (on the contrary to e2e tests, what can be launched from Linux/macOS/Windows in theory). We can add support for using Windows/macOS there if we cross-compile `vector` for linux - but that would significantly increas the build time and won't be fit for quick iterations anymore.
I want to address it at some point. But not now - and most likely after we ship the "MVP for the integration".

Author: Tried it on Windows - got stuck at the building vector into release artifacts. This should be easy to fix.
We just need to give Windows a little bit more attention to solve all the necessary dependencies before thinking about k8s-specific stuff.

Vector release artifacts are prepared for E2E tests, so the ability to do that
is required too, see Vector [docs](https://vector.dev) for more details.

##### Running the E2E tests

To run the E2E tests, use the following command:

```shell

> CONTAINER_IMAGE_REPO=<your name>/vector-test make test-e2e-kubernetes

Reviewer: Why not use `timberio/vector:dev` or something for a default so users don't need this?

Author: Users don't have push access to `timberio/vector:dev`, they'll have to specify something either way.
I was thinking about providing a default value if minikube cache is used, sane value in that case would be `localhost/<something>` due to some CRI internals.

Reviewer: That sounds reasonable. :)

Author: Added the default when `USE_MINIKUBE_CACHE=true` is set.

Reviewer: Maybe you missed it in the commit? 
![image](https://user-images.githubusercontent.com/130903/88424701-df1f7c00-cda2-11ea-8d51-0d147187329c.png)

Author: You still have to pass `USE_MINIKUBE_CACHE=true`.
We can actually autodetect whether `minikube` is used or now based on the `kubectl` context.
However, this would complicate the test setup a bit too much, and I chose not to do it. Besides technical reasons, the user experience won't change much, as k8s e2e tests still require a lot of involvement from the user. There was an attempt in eliminating this involvement before, with our previous k8s impl, but it wasn't very successful due to the amounts of edge cases. So, this time we took a different route and made things more robust, but less automatic.

`CONTRIBUTING.md` clearly documents the proper way of invoking the command, so this shouldn't be a problem for our users.

```

Where `CONTAINER_IMAGE_REPO` is the docker image repo name to use, without part
after the `:`. Replace `<your name>` with your Docker Hub username.

> This will run Kubernetes E2E tests suite. If the command exit with 0 status

Reviewer: I don't think this line is necessary. I think most users are able to tell that the output of a command has run and that it did not fail.

Author: True :smile: It's mostly here for completeness, but it is indeed redundant for the majority of the users.
I'll remove it, good find!

code - tests passed.

> You can also pass additional parameters to adjust the behavior of the test:

Reviewer: I suggest you namespace the k8s specific variables to like `KUBERNETES_USE_CACHE` and `KUBERNETES_CONTAINER_IMAGE_REPO` so we don't pollute the global environment namespace with variables whose names are relatively unclear.

Author: Prefixing all of them with `KUBERNETES_` is way too much characters to type, but `K8S_` would be fine I guess.
That said, I don't think global env pollution is an issue here. Unless the variables set globally at the system level (i.e. like in `~/.profile`) - but then it'd require more than just `K8S_` for them to make sense universally. I woundn't worry about that case.

Given that those are used in a very specific command, I don't think it's big issue. It's easy to type and remember though - and this is a big plus IMO.
If using them in `.envrc` or similar - it's often possible to use comments.

> * `QUICK_BUILD=true` - use development build and a skaffold image from the dev

Reviewer: How about `DEVELOPMENT=true`? I'd like to use that same env in other tasks so it'd be nice to keep the number of vars down.

Author: I'm pretty sure `DEVELOPMENT` didn't exist when I implemented this :smile: I'll see it if fits - we might want to deliberately have those different. I'll check and respond here.

Reviewer: It doesn't yet! :( You've probably noticed we use separate make jobs for this distinction. Maybe that's appropriate here too?

Author: You're thinking about `make test-e2e-kubernetes` and a `make test-e2e-kubernetes-quick`, sth like that? I thought about doing it like that, but I discarded it because we'd still need to pass env vars is some cases, and adding entry points complicates things worse than just having one configurable one imo. It's a good idea though, thx for bringing this up.

I understand what you mean with `DEVELOPMENT` now. Well, I wouldn't worry about it now. Let's streamline it when it becomes a problem. For now, `QUICK_BUILD` captures the distinction with the regular flow better (for the "how do I make it quick!!1" situations :smile: ), and with lack of widely used `DEVELOPMENT` var - I think it'd be easier to find.

Reviewer: I do not think `QUICK_BUILD` is a good name for this value.

  flow instead of a production docker image. Significantly speeds up the
  preparation process, but doesn't guarantee the correctness in the release
  build. Useful for development of the tests or Vector code to speed up the
  iteration cycles.

* `USE_MINIKUBE_CACHE=true` - instead of pushing the built docker image to the
  registry under the specified name, directly load the image into
  a `minikube`-controlled cluster node.
  Requires you to test against a `minikube` cluster. Eliminates the need to have
  a registry to run tests.

> * `CONTAINER_IMAGE_REPO=<your name>/vector-test:tag` - completely skip the step

Reviewer: I guess this exists because there is no `make build-docker-....` images and the `release` tasks are kind of awkward?

Author: Oops, there's a mistake. This is supposed to be `CONTAINER_IMAGE`, not `CONTAINER_IMAGE_REPO`. Hope it makes more sense.
This can be set if you want to bypass image building altogether, and use a completely custom image. This might be useful, for instance, for testing release images. I used it for experimenting with vector building, and to avoid rebuilding vector each time I change tests. Overall, it's very useful to have such option. Sorry for an error - this is definitely confusing as is.

Author: Also, if the `CONTAINER_IMAGE` is not set - a new image is built using the `CONTAINER_IMAGE_REPO` and a unique random tag. This process uses `make` (unless `QUICK_BUILD` is requested).

To cope with the complexity and ensure we maintain high quality, we use
E2E (end-to-end) tests.

> E2E tests normally run at CI, so there's typically no need to run them
> manually.

##### Requirements

* `kubernetes` cluster (`minikube` has special support, but any cluster should
  work)
* `docker`
* `kubectl`
* `bash`

Vector release artifacts are prepared for E2E tests, so the ability to do that
is required too, see Vector [docs](https://vector.dev) for more details.

##### Running the E2E tests

To run the E2E tests, use the following command:

```shell
CONTAINER_IMAGE_REPO=<your name>/vector-test make test-e2e-kubernetes
```

Where `CONTAINER_IMAGE_REPO` is the docker image repo name to use, without part
after the `:`. Replace `<your name>` with your Docker Hub username.

You can also pass additional parameters to adjust the behavior of the test:

* `QUICK_BUILD=true` - use development build and a skaffold image from the dev
  flow instead of a production docker image. Significantly speeds up the
  preparation process, but doesn't guarantee the correctness in the release
  build. Useful for development of the tests or Vector code to speed up the
  iteration cycles.

> * `USE_MINIKUBE_CACHE=true` - instead of pushing the built docker image to the

Reviewer: So by default when someone runs `make test-e2e-kubernetes CONTAINER_IMAGE_REPO=foo/bar` that pushes to the Docker hub? And this is the default?

Author: Yep. This shouldn't be surprising. Delivering images via registry is a standard way to make an image available to the Kubernetes. There's simply no other generic way of doing it. `minikube` has special support for delivering images into a cluster it manages without the registry, but we don't want to reimplement all that logic for a generic k8s cluster.
Users can spin up local registry if they don't want to use a cloud one.
